src/QueryEditor.tsx

```diff
@@ -1,47 +1,74 @@
 import React from 'react';
-import { css } from '@emotion/css';
 import { useAsync } from 'react-use';
-import { CoreApp, QueryEditorProps, SelectableValue } from '@grafana/data';
+import { QueryEditorProps, SelectableValue } from '@grafana/data';
 import { DataFrameDataSource } from './datasource';
-import { Column, DataframeQuery, QueryColumn } from './types';
-import { Button, InlineField, Input, MultiSelect } from '@grafana/ui';
+import { Column, DataframeQuery, isValidQuery, QueryColumn } from './types';
+import { InlineField, InlineSwitch, Input, MultiSelect, Select } from '@grafana/ui';
+import { decimationMethods, defaultDecimationMethod } from './constants';
 
 type Props = QueryEditorProps<DataFrameDataSource, DataframeQuery>;
 
-export const QueryEditor: React.FC<Props> = ({ app, query, datasource, onChange, onRunQuery }) => {
+export const QueryEditor: React.FC<Props> = ({ query, datasource, onChange, onRunQuery }) => {
   const tableMetadata = useAsync(async () => {
     return query.tableId ? await datasource.getTableMetadata(query.tableId) : null;
   }, [query.tableId]);
 
+  const runQueryIfValid = () => isValidQuery(query) && onRunQuery();
+
+  const handleQueryChange = (value: DataframeQuery, runQuery?: boolean) => {
+    onChange(value);
+    if (runQuery) {
+      runQueryIfValid();
+    }
+  };
+
   const handleIdInputBlur = (e: React.FocusEvent<HTMLInputElement>) => {
     if (query.tableId !== e.currentTarget.value) {
-      onChange({ ...query, tableId: e.currentTarget.value, columns: [] });
+      handleQueryChange({ ...query, tableId: e.currentTarget.value, columns: [] });
     }
   };
 
   const handleColumnChange = (items: Array<SelectableValue<string>>) => {
     const columns = items.map(({ value, dataType, columnType }) => ({ name: value!, dataType, columnType }));
-    onChange({ ...query, columns });
+    handleQueryChange({ ...query, columns });
   };
 
   return (
     <>
       <InlineField label="Id" error="Table does not exist" invalid={!!tableMetadata.error}>
         <Input defaultValue={query.tableId} width={26} onBlur={handleIdInputBlur} />
       </InlineField>
-      <InlineField label="Columns">
+      <InlineField label="Columns" tooltip="The columns to include in the response data.">
         <MultiSelect
           isLoading={tableMetadata.loading}
           options={columnsToOptions(tableMetadata.value?.columns)}
           onChange={handleColumnChange}
+          onBlur={runQueryIfValid}
```

Comment:
We don't want to re-run the query multiple times while the user is adding columns, so we only run it when the dropdown loses focus.